sherpa/ui/tests/test_ui_plot.py

```diff
@@ -2952,3 +2952,150 @@ def test_get_foo_component_plot_recalc(session, label, idval):
+
+
+@pytest.mark.parametrize("session", [BaseSession, AstroSession])
+@pytest.mark.parametrize("ptype", ["source", "model"])
+def test_plot_xxx_components_simple_mpl(session, ptype, requires_pylab):
+    """Simple check of plot_xxx_components using matpotlib"""
+
+    from matplotlib import pyplot as plt
+
+    s = session()
+    s._add_model_types(basic)
+
+    s.set_default_id(3)
+    s.load_arrays(3, [1, 10, 50], [2, 3, 4])
+
+    c1 = s.create_model_component("scale1d", "c1")
+    c2 = s.create_model_component("const1d", "c2")
+    c3 = s.create_model_component("box1d", "c3")
+
+    c1.c0 = 0.5
+    c2.c0 = 2
+    c3.ampl = 4
+    c3.xhi = 100
+    c3.xlow = 7
+    s.set_source(c1 * (c2 + c3))
+
+    pfunc = getattr(s, f"plot_{ptype}_components")
+    pfunc()
+
+    fig = plt.gcf()
+    assert len(fig.axes) == 1
+
+    axes = fig.axes[0]
+    assert axes.get_xlabel() == 'x'
+    assert axes.get_ylabel() == 'y'
+    assert axes.get_title() == 'Component plot'
+
+    assert len(axes.lines) == 2
+
+    assert axes.lines[0].get_xdata() == pytest.approx([1, 10, 50])
+    assert axes.lines[1].get_xdata() == pytest.approx([1, 10, 50])
+
+    # This should be c1 * c2 and c1 * c3
+    assert axes.lines[0].get_ydata() == pytest.approx([1, 1, 1])
+    assert axes.lines[1].get_ydata() == pytest.approx([0, 2, 2])
+
+
+@pytest.mark.parametrize("session", [BaseSession, AstroSession])
+@pytest.mark.parametrize("ptype", ["source", "model"])
+def test_plot_xxx_components_simple_bokeh(session, ptype):
+    """Simple check of plot_xxx_components using bokeh"""
+
+    backend_mod = pytest.importorskip("sherpa.plot.bokeh_backend")
+    backend = backend_mod.BokehBackend()
+
+    s = session()
+    s._add_model_types(basic)
+
+    s.set_plot_backend(backend)
+
+    s.set_default_id(3)
+    s.load_arrays(3, [1, 10, 50], [2, 3, 4])
+
+    c1 = s.create_model_component("scale1d", "c1")
+    c2 = s.create_model_component("const1d", "c2")
+    c3 = s.create_model_component("box1d", "c3")
+
+    c1.c0 = 0.5
+    c2.c0 = 2
+    c3.ampl = 4
+    c3.xhi = 100
+    c3.xlow = 7
+    s.set_source(c1 * (c2 + c3))
+
+    pfunc = getattr(s, f"plot_{ptype}_components")
+    pfunc()
+
+    # What to check? At the moment *very* basic (less than
+    # test_plot_xxx_components_simple_mpl).
+    #
+    fig = backend.current_fig
+    assert len(fig.children) == 1
+
+    fig0 = fig.children[0][0]
+    assert len(fig0.axis) == 2
+    assert fig0.xaxis.axis_label == "x"
+    assert fig0.yaxis.axis_label == "y"
+    assert fig0.title.text == "Component plot"
+
+    """
+    # How would we check the plotted data?
+
+    assert axes.lines[0].get_xdata() == pytest.approx([1, 10, 50])
+    assert axes.lines[1].get_xdata() == pytest.approx([1, 10, 50])
+
+    # This should be c1 * c2 and c1 * c3
+    assert axes.lines[0].get_ydata() == pytest.approx([1, 1, 1])
+    assert axes.lines[1].get_ydata() == pytest.approx([0, 2, 2])
+    """
```

Comment:
```suggestion
    json_item = pytest.importorskip("bokeh.embed.json_item")
    # Don't bother walking the hirachy, just check it's in there
    'Component plot' in str(jout)
    """
    # How would we check the plotted data?

    assert axes.lines[0].get_xdata() == pytest.approx([1, 10, 50])
    assert axes.lines[1].get_xdata() == pytest.approx([1, 10, 50])

    # This should be c1 * c2 and c1 * c3
    assert axes.lines[0].get_ydata() == pytest.approx([1, 1, 1])
    assert axes.lines[1].get_ydata() == pytest.approx([0, 2, 2])
    """
```

Reply:
In principle, we can get the json representation that is sent to the browser and than contains any and all information in the plot, e.g.
```Python
jout = json_item(backend.current_fig)
```
and we can then traverse (or search) through the tree to get to to the data:
```Python
entries = jout['doc']['roots'][0]['attributes']['children'][0][0]['attributes']['renderers'][1]['attributes']['data_source']['attributes']['data']['entries']
entries
[('x',
  {'type': 'ndarray',
   'array': {'type': 'bytes', 'data': 'AQAAAAoAAAAyAAAA'},
   'shape': [3],
   'dtype': 'int32',
   'order': 'little'}),
 ('y',
  {'type': 'ndarray',
   'array': {'type': 'bytes', 'data': 'AAAAAAAAAAAAAAAAAAAAQAAAAAAAAABA'},
   'shape': [3],
   'dtype': 'float64',
   'order': 'little'})]
```
Unfortunately, however, I have not found a way to get the numpy array back fromthe binary encoding. It's clearly possible in theory, I just don't know how (I tried, but the buffer has more bytes than needed to encode the shape*dtype array, so I assume that there is some error checking byte added or something like that. There is a spec for that and there obviously is code within bokeh to encode it, so I could go and find that, but at this point I'm afraid that it gets too fragile to be useful, e.g. what if the details change between bokeh versions?).

We could bypass the numeric comparison by just checking that the strings and included and assume that if they are, it's probably correct:
```python
assert 'AAAAAAAAAAAAAAAAAAAAQAAAAAAAAABA' in str(jout) 
```

That's easier to do and easier to update if it changes. However, I'm not sure it's worth it. If the plot is there, with the right title etc. I don't think the numbers are likely to be screwed up. Instead, we might do some simpler checks and see that there are actually two separate plots:
```python
assert 2 == len(jout['doc']['roots'][0]['attributes']['children'][0][0]['attributes']['renderers'])
```
but again, that's pretty deep in the hirachy and I don't think it's a good idea to tie ourselves to the bokeh datamodel that deep.

Reply:
I assume line 3094 should be

    assert 'Component plot' in str(jout)

?

I also agree that we don't want to  dive deep into the internal data representation of bokeh. I just know you'd said we could do this, so it's a bit sad we can't ;-)

Reply:
yes, you are correct.

About checking: We can do that, but it's harder than I thought. If we want that, I'll check to bokeh code and write a method to walk the json representation to find it even if it's a bit deeper in the hierarchy. I'm just not sure it's worth it.

Reply:
For this case I'm just going to drop it since I don't think that

    embed = pytest.importorskip("bokeh.embed")
    jout = embed.json_item(fig)
    assert 'Component plot' in str(jout)

really buys us anything (and trying to select the element in the JSON embedding directly is, as we've mentioned, requires too much intrenal knowledge of bokeh that may change over time).